### PR TITLE
Enable Master/Slaves Option

### PR DESCRIPTION
**Pull request by jacobbloomCDD**
Provides for tcp communication between a master and slave machines for faster testing.
All reports go back to the Master for analysis.  Any missing files from disconnects are rerun at the end.  If somehow there are still missing files, appropriate error reporting occurs.
Uses Unix sockets instead of TCP for master machine communication.

### CODE REVIEW

File: lib/rspec/multiprocess_runner/file_coordinator.rb [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      begin
+        @slave_socket.puts [COMMAND_FILE].to_json
+        file = @slave_socket.gets.chomp
+        if @spec_files_reference.include? file
+          return file
+        else
+          return nil # Malformed response, assume done, cease function
+        end
+      rescue
+        return nil # If Error, assume done, cease function
```
Comment (kellyjosephprice) [inline lib/rspec/multiprocess_runner/file_coordinator.rb]: I would normally suggest refactoring implicit returns, but I think being explicit here, especially with comments, is good. 👍

File: lib/rspec/multiprocess_runner/worker.rb [hunk 2]
```diff
@@ -19,6 +19,22 @@ module RSpec::MultiprocessRunner
   #   actually run specs.
   #
   # @private
+  class MockWorker
```
Comment (kellyjosephprice) [inline lib/rspec/multiprocess_runner/worker.rb]: The preceding rdoc is now pointing to the wrong class.

File: lib/rspec/multiprocess_runner/coordinator.rb [hunk 3]
```diff
@@ -36,7 +42,7 @@ def exit_code
       exit_code = 0
       exit_code |= 1 if any_example_failed?
       exit_code |= 2 if !failed_workers.empty?
-      exit_code |= 4 if !@spec_files.empty?
+      exit_code |= 4 if work_left_to_do? || @file_coordinator.missing_files.any?
```
Comment (kellyjosephprice) [inline lib/rspec/multiprocess_runner/coordinator.rb]: Should this be a new error code? Couldn't one be a process problem and the other a network problem?

File: lib/rspec/multiprocess_runner/file_coordinator.rb [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        count = 100
+        while @slave_socket.nil? do
+          begin
+            @slave_socket = TCPSocket.new @hostname, @port
+          rescue
+            raise if count < 0
+            count -= 1
+            sleep(6)
+          end
+        end
```
Comment (jacobbloomCDD) [inline lib/rspec/multiprocess_runner/file_coordinator.rb]: Sometimes it is doing a git pull before, which can cause a bit of time delay.  This ideally would be configurable, but I was getting tired of configuring everything.

File: lib/rspec/multiprocess_runner/command_line_options.rb [hunk 5]
```diff
@@ -119,6 +123,22 @@ def build_parser
           self.use_given_order = true
         end
 
+        parser.on("-p", "--port PORT", Integer, "Communicate using port (#{print_default port})") do |port|
+          self.port = port
+        end
+
+        parser.on("-H", "--hostname HOSTNAME", "Hostname for master machines (#{print_default hostname})") do |hostname|
```
Comment (kdole) [inline lib/rspec/multiprocess_runner/command_line_options.rb]: Maybe "Hostname of the master"?

**Comment by kdole** [comment]
I'm guessing the readme should be updated to describe the new multi-server architecture?

**Comment by jacobbloomCDD** [comment]
will do

File: lib/rspec/multiprocess_runner/coordinator.rb [hunk 6]
```diff
@@ -11,18 +12,23 @@ def initialize(worker_count, files, options={})
...
     end
 
     def run
       @start_time = Time.now
       expected_worker_numbers.each do |n|
         create_and_start_worker_if_necessary(n)
       end
-      run_loop
-      quit_all_workers
+      2.times do # rerun files missing from disconnects
```
Comment (kdole) [inline lib/rspec/multiprocess_runner/coordinator.rb]: Like we discussed, it would be nice to rewrite this to clarify what is happening.

**Comment by kellyjosephprice** [comment]
Changes LGTM.

**Pull request merged**